python/lsst/ts/audiotrigger/read_serial_temp_scanner.py

```diff
@@ -0,0 +1,313 @@
+        await super().start(kwargs=kwargs)
+
+    async def write_if_connected(self, string_to_write):
+        """Write message if connected."""
+        if self.simulation_mode:
+            return
+        if self.connected:
+            await self.write_str(string_to_write)
+
+    async def set_fan(self, setting):
+        """Set the fan on/off.
+
+        Parameters
+        ----------
+        setting : `int`
+            Turn the fan on or off.
+        """
+        if not self.configured or not self.pi.connected:
+            await self.write_if_connected(
+                "TS: Error: Not configured properly before actuating fan"
+            )
+            raise ValueError("Not configured properly before actuating fan")
+        await self.loop.run_in_executor(
+            None, functools.partial(self.pi.write, self.fan_gpio, setting)
+        )
+
+    async def set_fan_on(self):
+        """Set the fan on."""
+        await self.set_fan(Fan.ON)
+
+    async def set_fan_off(self):
+        """Set the fan off."""
+        await self.set_fan(Fan.OFF)
+
+    async def handle_data(self, new_data):
+        """Handle the incoming data.
+
+        Parameters
+        ----------
+        new_data : `dict`
+            New data.
+        """
+        # We only care about one sensor's reading for operating the fan
+        try:
+            new_data[self.fan_sensor] = float(new_data[self.fan_sensor])
+        except Exception as e:
+            await self.write_if_connected(
+                f"TS: Exception trying to convert data to int... Data: {new_data[self.fan_sensor], {str(e)}}"
+            )
+            self.log.exception(
+                f"Exception trying to convert data to int... Data: {new_data[self.fan_sensor], {str(e)}}"
+            )
+            return
+        if new_data[self.fan_sensor] >= self.fan_turn_on_temp:
+            self.log.info(f"Turning ON fan, temperature: {new_data[self.fan_sensor]}")
+            await self.set_fan_on()
+        elif new_data[self.fan_sensor] < self.fan_turn_off_temp:
+            self.log.info(f"Turning OFF fan, temperature: {new_data[self.fan_sensor]}")
+            await self.set_fan_off()
+
+        # Now for telemetry, do a rolling average of all 8 sensors
+        new_temperature = 0
+        for reading in new_data:
+            data = float(new_data[reading])
+            self.log.info(f"reading: {data}")
+            new_temperature += data
+        new_temperature = new_temperature / len(new_data)
+
+        # Handle initial data loading
+        if self.first_run:
+            self.first_run = False
+            for i in range(len(self.rolling_temperature)):
+                self.rolling_temperature[i] = new_temperature
+        else:
+            # Update rolling temperature windows
+            # freshest data starts at 0 index and goes towards max length index
+            for i in range(len(self.rolling_temperature) - 1, 0, -1):
```

Comment:
I assume this is Joseph's code but I think this may not have the intended effect. As far as I know `self.rolling_temperature[len(self.rolling_temperature) - 1]` and `self.rolling_temperature[-1]` are the same element in the array. What is the rolling window needed for?

Reply:
The rolling window appears to be for setting new average temperatures calculated from the latest channel streams.